# Taxa Interna de Retorno

A taxa interna de retorno (Intern Rate of Return(IRR)) ou TIR, é um dos indicadores de análise de investimentos amplamente utilizado na engenharia econômica. Seu cálculo consiste em mensurar a taxa de retorno do investimento escolhido. É determinada pela realização de projeções de fluxos de caixa a valor presente considerando o VPL(Valor Presente Líquido) igual a zero. 

![TIR](https://www.voitto.com.br/assets/ckeditor2/plugins/imageuploader/uploads/941ca2fa0.png)

A TIR é o percentual de retorno do investimento em análise. Logo, o quanto maior seu valor, maior a possibilidade do projeto ser aprovado. 

Em análises de investimentos, a TIR é comparada com a TMA(Taxa Mínima de Atratividade). Como regra, a TIR é avaliada seguindo as seguintes premissas:

1. Se **TIR** > **TMA**, o projeto deve ser aprovado.
2. Se **TIR** < **TMA**, o projeto deve ser rejeitado.
3. Se **TIR** = **TMA**, a avaliação deverá utilizar outra ferramenta de análise para corroborar a decisão.

Como vantagem de sua aplicação está seu valor relativo(percentual), garantindo maior facilidade de interpretação. Além disso, considera o valor do dinheiro ao longo do tempo. 



**Exemplo 1**: Um investimento inicial de R$ 500,00 será realizado. Os fluxos de caixas serão realizadas ao longo de 8 anos na quantia de 150,00 a taxa de 10%a.a. Qual a taxa interna de retorno ?

In [ ]:
!pip install numpy_financial 

In [ ]:
#utilizando a lib numpy_financial que contém um método específico para cálculo da TIR

import numpy_financial as npf

In [ ]:
#fluxo de caixa

values= [-500, 150, 150, 150, 150, 150, 150, 150, 150]

In [ ]:
irr = npf.irr(values) * 100

In [ ]:
round(irr,2)

24.95

A TIR apresenta desvantagem de considerar valores constantes de fluxos de caixa ao longo do tempo. Além disso, não pondera os investimentos com suas possíveis taxas de reinvestimentos.

Para resolução destes problemas, deve-se utilizar a *Taxa Interna de Retorno Modificada* - **TIRM**. 



**Exemplo 1**: Considerando o fluxo de caixa: [ -150000, 40000, 40000, 45000, 45000, 50000 ], taxa de reinvestimento de 9% e custo de oportunidade de 12%, informe a TIRM.


In [ ]:
values2= [-150000, 40000, 40000, 45000, 45000, 50000]

In [ ]:
tirm = npf.mirr(values2, 12, 0.09)

In [ ]:
round(tirm*100, 2)

11.7

Como MTIR < Custo de Oportunidade, o projeto não é viável.

### Conflito TIR x VPL

Vamos analisar os seguinte fluxos de caixa. 

In [ ]:
import pandas as pd

In [ ]:
data = pd.DataFrame({'A':[-500, 200, 250, 300, 350], 
                     'B': [-1000, 450, 450, 450, 450]})

In [ ]:
data

A     B
0 -500 -1000
1  200   450
2  250   450
3  300   450
4  350   450

In [ ]:
#VPL - custo de oportunidade = 12%

def vpl(cashflow, rate):
  npv =[]

  for i in cashflow.index:
    value = cashflow[i] / ((1+rate)**i)

    npv.append(value)
  
  return round(sum(npv),2)

In [ ]:
#VPLA - A
VPL_A = vpl(data['A'], 0.12)
VPL_A

313.84

In [ ]:
#VPL - B
VPL_B = vpl(data['B'], 0.12)
VPL_B

366.81

In [ ]:
#TIR - A 

TIR_A = round(npf.irr(data['A']),4)*100

TIR_A

36.44

In [ ]:
#TIR - B

TIR_B = round(npf.irr(data['B']),4)*100

TIR_B

28.49

Apresentando os indicadores de análise dos fluxos de caixas acima, identifica-se que o VPL de B  > VPL de A, mas a TIR de B < TIR de A. 

Logo, qual projeto aprovar ?

Para casos como esse utilizaremos o **fluxo incremental de fisher** que avalia o fluxo de caixa da diferença dos fluxos apresentados.

In [ ]:
for i in data.index: 
  data['B-A'].iloc[i] =  int(data['B'].iloc[i]) - int(data['A'].iloc[i])

In [ ]:
data

A     B  B-A
0 -500 -1000 -500
1  200   450  250
2  250   450  200
3  300   450  150
4  350   450  100

In [ ]:
VPLincremental = round(VPL_B - VPL_A,2)

VPLincremental

52.97

In [ ]:
TIRincremental = round(npf.irr(data['B-A']),4)*100

TIRincremental

17.8

In [ ]:
analise = pd.DataFrame({'':['Investimento', 'VPL', 'TIR'],
                        
                        'A': [int(data['A'].iloc[0]),
                              VPL_A,
                              TIR_A],
                        
                        'B': [int(data['B'].iloc[0]),
                              VPL_B,
                              TIR_B],
                        
                         'B-A': [int(data['B-A'].iloc[0]),
                              VPLincremental,
                              TIRincremental]
                        })

In [ ]:
analise

A        B     B-A
0  Investimento -500.00 -1000.00 -500.00
1           VPL  313.84   366.81   52.97
2           TIR   36.44    28.49   17.80

Podemos dizer que o projeto B apresenta maior valor real, até uma taxa de 17.8%,. A partir dessa taxa, é interessante investir no projeto A.